hack/install-vm.sh

```diff
@@ -141,26 +149,261 @@ EOF
+#!/bin/bash
+
+AGENT_CONF_PATH="/etc/antrea"
+AGENT_LOG_DIR="/var/log/antrea"
+
+declare -a OVS_KERNEL_MODULES=("vport_geneve" "openvswitch")
+declare -a DOCKER_SERVICE=("ovs" "antrea-agent")
+
+load_ovs_modules() {
+    echo "Loading OVS kernel"
+    for module in "${OVS_KERNEL_MODULES[@]}"; do
+        lsmod | grep -q $module
+        if [[ $? -ne 0 ]]; then
+            echo "Load $module kernel module"
+            modprobe $module
+        fi
+    done
+}
+
+exec > >(tee $AGENT_LOG_DIR/vm-watchdog.log) 2>&1
+
+docker info
+if [[ $? -ne 0 ]]; then
+    echo "Starting docker service"
+    systemctl start docker
+    sleep 2
+fi
+
+load_ovs_modules
+
+while(true); do
+    for service in "${DOCKER_SERVICE[@]}"; do
+        if [ ! "$(docker ps -q -f name=$service)" ]; then
+            if [ "$(docker ps -aq -f status=exited -f name=$service)" ]; then
+                echo "$service is in exited state"
+            fi
+            echo "Executing docker-compose up"
+            docker-compose -f $AGENT_CONF_PATH/docker-compose.yaml up -d
+        fi
+    done
+    sleep 5
+done
+EOF
+}
+
+# Create watchdog service as a systemd service for monitoring/creating antrea specfic containers/services
+create_antrea_agent_watchdog_service() {
+    DESCRIPTION="Create antrea-agent watchdog service to monitor OVS and antrea-agent containers"
+    SERVICE_NAME="antrea-agent-watchdog.service"
+    IS_ACTIVE=$(sudo systemctl is-active $SERVICE_NAME)
+    if [ "$IS_ACTIVE" == "active" ]; then
+        # restart the service
+        echo "Systemd service $SERVICE_NAME is running"
+        echo "Restarting $SERVICE_NAME service"
+        sudo systemctl restart $SERVICE_NAME
+    else
+        # create service file
+        echo "Creating systemd service file for $SERVICE_NAME"
+        sudo cat > /etc/systemd/system/${SERVICE_NAME} << EOF
+[Unit]
+Description=$DESCRIPTION
+After=network.target
+[Service]
+ExecStart=$AGENT_CONF_PATH/antrea-agent-watchdog.sh
+Restart=on-failure
+[Install]
+WantedBy=multi-user.target
+EOF
+        # restart daemon, enable and start service
+        echo "Reloading daemon and enabling service $SERVICE_NAME"
+        sudo systemctl daemon-reload
+        sudo systemctl enable ${SERVICE_NAME//'.service'/} # remove the extension
+        sudo systemctl start ${SERVICE_NAME//'.service'/}
+        if [ $? -eq 0 ]; then
+            echo "Antrea-agent watchdog service started successfully"
+        else
+            echoerr "ERROR: Antrea-agent watchdog service could not be started, error $status"
+            exit 1
+        fi
+    fi
+}
+
+# Function to deploy and run antrea specfic containers
+deploy_antrea_containers() {
+    set +eo pipefail
+    echo "Extract Docker image from tar file"
+    docker pull $ANTREA_DOCKER_IMAGE
+    if [[ "$?" -ne 0 ]]; then
+        echoerr "ERROR!!docker pull from tar file failed"
+        exit 2
+    fi
+    # Load ovs kernel modules
+    load_ovs_modules
```

Comment:
Instead of calling load_ovs_modules,  docker compose here and starting the containers, can we rely on  watchdog service to deploy and run the containers?

This would avoid code duplication in install-vm.sh and antrea-agent-watchdog.sh.